Power control (and some refactoring)

kbembedded opened the pull request:
TODO: Need to add options to cmdline to set specific power states at startup

Also, @FryDay, would you be interested in adding some more graphics like you did for the LED? Basically a battery indicator like you'd see on a smart device: white battery outline, two bars on the inside, both for high, one for low, none (and maybe change the battery outline to red?) for depleted. And some kind of AC/external power input indicator. I'm not sure the best way to showcase that. If you don't have any ideas I can dig around some more. Also, feel free to comment on changes, it looks like you arn't a part of this project normally so I can't request a review.

kbembedded committed:
msemu: Add power control

Allows for setting the status of the battery and AC adapter during
runtime via keyboard shortcuts.

In order to facilitate this, some other changes were made to make
some runtime operations more logical. The ms_ctx struct has added
variables and comments to denote what is "hardware." These are now
accessed from IO reads where applicable. e.g. for the power
control, setting the battery level modfies ms_ctx, then reads of
the IO port return data based on the current ms_ctx. This was
done this way since io_init() writes all 0x00 to each port (for
now), and the firmware can also end up writing to these IO ports
even though they never really should since they are input-only.
This made it difficult to maintain the power state or other
"hardware" states in IO ports directly when they could change
without the "hardware" emulation knowing about it. Writes to
these IO ports go to the IO buffer but do not affect read.

This also removed some unused ms_ctx struct members.

kbembedded committed:
lcd: Minor refactor

Reduces function args for all LCD functions, makes calling
cleaner. Also touched a few style inconsistencies. At some point
the LCD system could use a deeper cleanup.

Added lcd_deinit()

kbembedded committed:
io: Minor refactor

Pass around whole struct rather than having callers worry about
the correct notation.

kbembedded committed:
mem: Minor refactor

Moved functions to all use ms_ctx rather than passing around bufs
like previous commits.

Moved the RAM image buffer to a buffer in ms_ctx

kbembedded commented on src/msemu.c:
```diff
@@ -115,7 +115,8 @@ void ms_power_on_reset(ms_ctx *ms)
 	lcd_init(&ms->lcd_dat1bit, &ms->lcd_datRGBA8888, &ms->lcd_cas);
 	ram_init(&ms->ram, NULL);
-	io_init(&ms->io);
+	//io_init(&ms->io);
```
Whoops, didn't mean to leave this commented out.

kbembedded commented on src/msemu.c:
```diff
@@ -382,7 +384,19 @@ Z80EX_BYTE z80ex_pread (
-		ret = (uint8_t)0xE0 | ((~ms->power_button & 1) << 4);
+		tmp = 0;
+		if (ms->ac_status) tmp |= MISC9_AC_GOOD;
+		// XXX: Make the 4 a define
```
Forgot about this...

kbembedded commented on src/msemu.c:
```diff
@@ -451,7 +465,7 @@ void z80ex_pwrite (
 	void *user_data)
 {
-
+	uint8_t tmp;
```
This can go

kbembedded commented on src/msemu.c:
```diff
@@ -568,6 +582,55 @@ int process_interrupts (ms_ctx* ms)
 
+/* Enable, disable, or toggle AC adapter status
+ *
+ * Writes to the relevant IO port
```
Doesn't actually write to IO, fix comment

kbembedded commented on src/msemu.c:
```diff
@@ -568,6 +582,55 @@ int process_interrupts (ms_ctx* ms)
+
+/* Set battery voltage status, high, low, depleted
+ *
+ * Writes to the relevant IO port
```
Fix comment, doesn't write to IO port

kbembedded commented on src/msemu.c:
```diff
@@ -583,7 +646,8 @@ int ms_init(ms_ctx* ms, ms_opts* options)
 
 	ms->interrupt_mask = 0;
-	ms->power_button = 0;
+	// XXX: Should this be abstracted? Seems silly to do so.
```
Figure out what to do with this

kbembedded commented on src/msemu.c:
```diff
@@ -600,6 +664,8 @@ int ms_init(ms_ctx* ms, ms_opts* options)
 	if (lcd_init(&ms->lcd_dat1bit, &ms->lcd_datRGBA8888, &ms->lcd_cas))
 		return MS_ERR;
 	if (io_init(&ms->io)) return MS_ERR;
+	ms_power_ac_set_status(ms, AC_GOOD);
```
Change these to be evaluated from command line, default to AC_GOOD and BATT_HIGH though

kbembedded commented on src/mem.c:
```diff
@@ -290,32 +290,32 @@ int cf_init(uint8_t **cf_buf, ms_opts *options)
+int cf_deinit(ms_ctx *ms, ms_opts *options)
 {
-	assert(*cf_buf != NULL);
+	assert(ms->cf != NULL);
```
Remove this, waste of cpu time

FryDay commented:
@kbembedded I'd be happy to work on some more graphics. I'll let you know what kind of ideas I come up with.

FryDay commented:
#64 implements the new indicators for the battery and ac status.

kbembedded commented:
This should all be good to go!

FryDay committed:
Power UI (#64)

Add battery and power indicator icons

FryDay committed:
New splash screen

kbembedded committed:
main: Add hotkey description to help output

kbembedded committed:
msemu: Cleaned up power on/off/reset

Cleans up F12 checking in ui, F12 is now checked and set from
the main loop. After each change it calls a ms_power_hint().
This function was added to allow the ui to hint at a state change.
The hint function checks that the power is currently off, and
that the power button was pressed, before calling ms_power_on_reset().

ram_init() was also moved from ms_power_on_reset() to ms_power_off().
This is so that a hard reset, with CTRL+r, is more akin to hitting
the physical reset button on a machine where RAM would stay intact.
This way, only a clean power off would re-initialize RAM.

This also has the side effect of better mimicing real hardware.
The MailStation reset button is connected to the wake signal. Now,
from the main ui splashscreen, hitting CTRL+r will correctly cause
the emulation to start up.
